_posts/2015-08-04-renaming-tables-with-zero-downtime.md

```diff
@@ -0,0 +1,80 @@
+Bringing our site down for several minutes to run the migrations and deploy the updated code was unacceptable so I started investigating to come up with a solution.
+
+In this example, we will be renaming `alice` to `bob`.
+
+# Using a View
+
+The first thing I came up with was using [MySQL Views] to alias the table so it would be accessible by both names (`alice` and `bob`).
+I came to find out that [many MySQL Views are updatable and insertable][updatable views]:
+
+> Some views are updatable and references to them can be used to specify tables to be updated in data change statements. That is, you can use them in statements such as UPDATE, DELETE, or INSERT to update the contents of the underlying table.
+
+This is super cool because it means that _all_ of the queries, reads and writes, can be done against the new name.
+Here we create a view with the new name that we want to end up with:
+
+{% highlight mysql %}
+CREATE VIEW `bob` AS SELECT * FROM `alice`;
+{% endhighlight %}
+
+# Updating Queries in Code
+
+Now that we have a view that acts just like a proper table, we can deploy the code that updates all queries to reference `bob` instead of `alice`.
+All actions, including writes, proxy through to the underlying table, `alice`.
+
+# Replacing the View With the Table
+
+The last piece of the puzzle is to rename the table and drop the old view.
+Fortunately, MySQL's [rename table] command allows you to atomically rename multiple tables:
+
+{% highlight mysql %}
+RENAME TABLE
+   `bob` TO `bob_old_view`,
+   `alice` to `bob`;
+
+DROP VIEW `bob_old_view`;
+{% endhighlight %}
+
+And just like that we have renamed the table without any loss of availability.
+
+# Caveats
+
+This approach has a few caveats that you should be aware of:
+
+## Broken Views
+
+While foreign keys _are_ updated when renaming the table, queries in views are _not_.
+In fact, immediately after the `RENAME TABLE` command in this example, the query backing `bob_old_view` is left untouched so it is still referencing `alice` which doesn't exist anymore.
+This doesn't matter in particular because we end up dropping the view in the next statement, but leaving a bad view does cause `mysqldump` to fail.
+
+## Missing Data in the Information Schema
+
+Also, we discovered one minor difference between the table and the view: the primary key for the view is not available in the `information_schema`.
+Our ORM uses this information to build its model of the object so it can generate queries to interact with the underlying table.
+Fortunately, we could easily provide the information it needed in code while the view was in use so it wouldn't have to look it up in the information schema table at all.
+
+There very well be more differences between tables and views that we didn't come across.
+
+# Conclusion
+
+All of the SQL commands were nearly instantaneous and worked flawlessly when used on a fairly large table with a modest number of reads and writes.
```

Comment:
That being said, maybe the likelihood of someone having a really similar use-case isn't very high. I think I like your second suggestion better than what's currently in the post.